aptos-move/aptos-gas/src/gas_meter.rs

```diff
@@ -257,6 +266,32 @@ impl AptosGasMeter {
         }
     }
 
+    #[inline]
+    fn charge_execution(&mut self, amount: InternalGas) -> PartialVMResult<()> {
+        self.charge(amount)?;
+
+        self.execution_gas_used += amount;
+        if self.feature_version >= 7
+            && self.execution_gas_used > self.gas_params.txn.max_execution_gas
+        {
+            Err(PartialVMError::new(StatusCode::EXECUTION_LIMIT_REACHED))
+        } else {
+            Ok(())
+        }
+    }
+
+    #[inline]
+    fn charge_io(&mut self, amount: InternalGas) -> PartialVMResult<()> {
+        self.charge(amount)?;
+
+        self.io_gas_used += amount;
+        if self.feature_version >= 7 && self.io_gas_used > self.gas_params.txn.max_io_gas {
```

Comment:
Can this break upgradability of aptos-framework package? We already have to add a mechanism to bypass tx size limit. Now that there's an io gas limit, how much headroom do we have before we need to bypass that limit as well?

Reply:
This is a great question but conceptually, I don't think this is going to break framework upgrades. Our goal here is to reduce transaction costs, not increasing them. Having separate limits for each individual limits will likely result in more transactions being allowed -- old transactions should be able to pass as long as we configure the new limits no stricter than the old one.